// src/pbx_smoke_test.js
import http from 'k6/http';
import { sleep, check } from 'k6';

export const options = {
    vus: 5, // 5 virtual users
    duration: '1m', // duration of the test is 1 minute
};

function makeRequestWithRetry(url, method, body = null, headers = {}, maxRetries = 3) {
    let res;
    let backoff = 5; // initial backoff time in seconds

    for (let i = 0; i < maxRetries; i++) {
        if (method === 'GET') {
            res = http.get(url, { headers });
        } else if (method === 'POST') {
            res = http.post(url, JSON.stringify(body), { headers });
        }

        if (res.status !== 429) {
            return res;
        }

        console.warn('Rate limit hit! Retrying after ' + backoff + 's');

        let jitter = Math.random() * 5;
        sleep(backoff + jitter);
        backoff *= 5;

    }

    return res;

}

export default function () {
    // let res = http.get('https://mypbx.telkom.co.id/user/login');
    // check(res, { "status is 200": (res) => res.status === 200 });
    sleep(1);
    

    const loginRequest = {
        email: 'superadmin@telkompbx.com',
        password: 'admin123',
        visitor_id: '243552f2e1c7629d66b0d3d4a78d7a2c',
    }
    const headersLoginPage = {
        'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8',
        'Content-Type': 'text/html; charset=utf-8',
    }

    const headers = {
        'Accept': 'application/json',
        'Content-Type': 'application/json',
    };

    const loginPage = makeRequestWithRetry('https://mypbx.telkom.co.id/user/login', 'GET', null, headersLoginPage);
    check(loginPage, {
        'User Successfully Access Login Page (200)': (r) => r.status === 200,
    }) || console.log('User Successfully Access Login Page!');

    const loginResponse = makeRequestWithRetry('https://mypbx.telkom.co.id/api/login', 'POST', loginRequest, headers);
    check(loginResponse, {
        'Login Success (200)': (r) => r.status === 200,
    }) || console.error('Failed to Login! Received status code: ' + loginResponse.status_text);

    if (loginResponse.status !== 200) {
        return;
    }

    const loginBodyResponse = loginResponse.json();

    console.log('Login Success! Token: ' + loginBodyResponse.data.token);
}

//     const profileHeaders = {
//         'Accept': 'application/json',
//         'Authorization': `Bearer ${loginBodyResponse.data.token}`,
//     };

//     const getProfile = makeRequestWithRetry('https://mypbx.telkom.co.id/api/profile', 'GET', null, profileHeaders);

//     check(getProfile, {
//         'Success Get Profile Status is 200': (r) => r.status === 200,
//         'Not subject to rate limit 429': (r) => r.status !== 429,
//     }) || console.error('Failed to Get User Profile! Received status code: ' + getProfile.status);

//     if (getProfile.status === 200) {
//         console.log('Get Profile Success! User Profile: ' + JSON.stringify(getProfile.json()));
//     }
// }
