// src/pbx_stress_test2.js
import http from 'k6/http';
import { sleep, check } from 'k6';
import { SharedArray } from 'k6/data';

export const options = {

    stages: [
        { duration: '1m', target: 100 },
        { duration: '2m', target: 200 },
        { duration: '5m', target: 300 },
        { duration: '10m', target: 500 },
        { duration: '2m', target: 0 },
    ],

    //   thresholds: {
    //     'http_req_failed': ['rate<0.01'], // error rate max 1%
    //     'http_req_duration': ['p(95)<1000'], // 95% response < 1s
    //   },
};

function getWithRetry(url, headers = {}, maxRetries = 5) {
    let res;
    let backoff = 1;

    for (let i = 0; i < maxRetries; i++) {
        res = http.get(url, { headers });

        if (res.status !== 0 && res.status !== 429 && res.status < 500) {
            return res;
        }

        console.warn(`[Retry ${i + 1}] ${url} failed with status ${res.status || 'NO_RESPONSE'}. Retrying in ${backoff}s`);
        sleep(backoff + Math.random() * 2);
        backoff = Math.min(backoff * 2, 10);
    }

    return res;
}

// Simulasi user credentials (bisa di-random nanti)
const users = new SharedArray('users', function () {
    return [
        { email: 'xyz@admin.com', password: 'admin#123' },
        // tambahkan user lain jika perlu
    ];
});

function login(user) {
    const loginUrl = 'https://mypbx.telkom.co.id/api/login';
    const payload = JSON.stringify({
        email: user.email,
        password: user.password,
        visitor_id: '243552f2e1c7629d66b0d3d4a78d7a2c',
    });

    const headers = {
        'Content-Type': 'application/json',
        'Accept': 'application/json',
    };

    let res = http.post(loginUrl, payload, { headers });

    check(res, {
        'Login Success (200)': (r) => r.status === 200,
    }) || console.error(`Login failed with status: ${res.status} - ${res.body}`);

    return res.status === 200 ? res.json().data.token : null;
}

export function setup() {
    const user = users[0]; // bisa pakai Math.random() untuk random user
    const token = login(user);
    if (!token) {
        throw new Error('Login gagal di setup');
    }
    return { token };
}

export default function (data) {
    const headers = {
        Authorization: `Bearer ${data.token}`,
        Accept: 'application/json',
    };

    //request ke dashboard
    const dashboardRes = getWithRetry('https://mypbx.telkom.co.id/app/show_dashboard', { headers });
    check(dashboardRes, {
        'Dashboard success (200)': (r) => r.status === 200,
    });

    sleep(1);
}
